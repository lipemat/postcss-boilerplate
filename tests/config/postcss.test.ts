const browserslist = require( 'browserslist' );
const postcssPresetEnv = require( 'postcss-preset-env' );

type Config = {
	browsers: Array<string>;
	features: {
		'focus-visible-pseudo-class': {
			replaceWith: string;
		};
	};
	processors: Array<{
		plugins?: Array<{
			postcssPlugin: string;
		}>
	}>;
	parser: string;
	map: boolean;
}

type GruntTask = {
	options: Config;
	files: {
		[ key: string ]: string;
	};
}

function getPostCSSConfig(): {
	toCSS: GruntTask;
	min: GruntTask;
	} {
	jest.resetModules();
	return require( '../../config/postcss.js' );
}

afterEach( () => {
	delete process.env.BROWSERSLIST;
} );


describe( 'postcss.js', () => {
	test( 'PostCSS config', () => {
		/**
		 * @notice If the browserslist results change, the snapshot will need to be updated.
		 */
		const config = require( '../../config/postcss' );
		expect( config ).toMatchSnapshot( 'develop' );
	} );


	test( 'sourceMap', () => {
		const config = require( '../../config/postcss' );
		expect( config.toCSS.options.map ).toEqual( true );
		expect( config.min.options.map ).toEqual( false );
	} );


	test( 'Browserslist config', () => {
		const expectedBrowsers = [ ...require( '@wordpress/browserslist-config' ) ];
		expectedBrowsers.push( 'not and_uc 15.5' );
		const creator = browsers => {
			return postcssPresetEnv( {
				browsers,
				features: {
					'focus-visible-pseudo-class': {
						replaceWith: ':global(.focus-visible)',
					},
				},
			} );
		};

		const config = getPostCSSConfig();
		// We want to make sure no matter what postcss-custom-properties is not included
		// if a user did not provided a custom browserslist to override.
		expect( config.toCSS.options.processors[ 3 ]?.plugins?.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 0 );

		expect( JSON.stringify( config.toCSS.options.processors[ 3 ] ) )
			.toEqual( JSON.stringify( creator( expectedBrowsers ) ) );
		expect( JSON.stringify( config.min.options.processors[ 3 ] ) )
			.toEqual( JSON.stringify( creator( expectedBrowsers ) ) );

		// and_uc 15.5 requires postcss-custom-properties.
		process.env.BROWSERSLIST = 'and_uc 15.5';
		const config2 = getPostCSSConfig();
		expect( config2.toCSS.options.processors[ 3 ]?.plugins?.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 1 );
		expect( JSON.stringify( config2.toCSS.options.processors[ 3 ] ) )
			.toEqual( JSON.stringify( creator( [ 'and_uc 15.5' ] ) ) );
		expect( JSON.stringify( config2.min.options.processors[ 3 ] ) )
			.toEqual( JSON.stringify( creator( [ 'and_uc 15.5' ] ) ) );


		// @notice If this fails, we can probably remove the override in favor of default wp.
		const wpDefaultBrowsers = [ ...require( '@wordpress/browserslist-config' ) ];
		process.env.BROWSERSLIST = browserslist( wpDefaultBrowsers );
		const config3 = getPostCSSConfig();
		expect( config3.toCSS.options.processors[ 3 ]?.plugins?.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 1 );
	} );
} );
