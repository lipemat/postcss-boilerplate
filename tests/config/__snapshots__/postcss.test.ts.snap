// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcss.js PostCSS config: develop 1`] = `
{
  "min": {
    "files": {},
    "options": {
      "map": false,
      "parser": {
        "parse": [Function],
        "stringify": [Function],
      },
      "processors": [
        {
          "Once": [Function],
          "postcssPlugin": "postcss-import",
        },
        [Function],
        [Function],
        {
          "plugins": [
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-color-function",
            },
            {
              "postcssPlugin": "postcss-normalize-display-values",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "postcssPlugin": "postcss-font-format-keywords",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-oklab-function",
            },
            {
              "postcssPlugin": "postcss-text-decoration-shorthand",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "Once": [Function],
              "postcssPlugin": "postcss-custom-media",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-image-set-function",
            },
            {
              "AtRule": {
                "custom-media": [Function],
                "media": [Function],
              },
              "postcssPlugin": "postcss-media-minmax",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-nesting",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-dir-pseudo-class",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-lab-function",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-blank-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-page-break",
            },
            {
              "Once": [Function],
              "postcssPlugin": "postcss-font-variant",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-has-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-stepped-value-functions",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-trigonometric-functions",
            },
            {
              "browsers": [
                "> 1%",
                "last 1 Android versions",
                "last 1 ChromeAndroid versions",
                "last 2 Chrome versions",
                "last 2 Firefox versions",
                "last 2 Safari versions",
                "last 2 iOS versions",
                "last 2 Edge versions",
                "last 2 Opera versions",
                "not and_uc 15.5",
              ],
              "info": [Function],
              "options": {
                "overrideBrowserslist": [
                  "> 1%",
                  "last 1 Android versions",
                  "last 1 ChromeAndroid versions",
                  "last 2 Chrome versions",
                  "last 2 Firefox versions",
                  "last 2 Safari versions",
                  "last 2 iOS versions",
                  "last 2 Edge versions",
                  "last 2 Opera versions",
                  "not and_uc 15.5",
                ],
              },
              "postcssPlugin": "autoprefixer",
              "prepare": [Function],
            },
            {
              "RuleExit": [Function],
              "postcssPlugin": "postcss-progressive-custom-properties",
            },
            {
              "OnceExit": [Function],
              "postcssPlugin": "postcss-preset-env",
            },
          ],
          "postcssPlugin": "postcss-preset-env",
        },
        [Function],
        [Function],
        {
          "OnceExit": [Function],
          "postcssPlugin": "postcss-hash",
        },
        {
          "OnceExit": [Function],
          "postcssPlugin": "clean",
        },
      ],
    },
  },
  "toCSS": {
    "files": {},
    "options": {
      "map": true,
      "parser": {
        "parse": [Function],
        "stringify": [Function],
      },
      "processors": [
        {
          "Once": [Function],
          "postcssPlugin": "postcss-import",
        },
        [Function],
        [Function],
        {
          "plugins": [
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-color-function",
            },
            {
              "postcssPlugin": "postcss-normalize-display-values",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "postcssPlugin": "postcss-font-format-keywords",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-oklab-function",
            },
            {
              "postcssPlugin": "postcss-text-decoration-shorthand",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "Once": [Function],
              "postcssPlugin": "postcss-custom-media",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-image-set-function",
            },
            {
              "AtRule": {
                "custom-media": [Function],
                "media": [Function],
              },
              "postcssPlugin": "postcss-media-minmax",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-nesting",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-dir-pseudo-class",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-lab-function",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-blank-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-page-break",
            },
            {
              "Once": [Function],
              "postcssPlugin": "postcss-font-variant",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-has-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-stepped-value-functions",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-trigonometric-functions",
            },
            {
              "browsers": [
                "> 1%",
                "last 1 Android versions",
                "last 1 ChromeAndroid versions",
                "last 2 Chrome versions",
                "last 2 Firefox versions",
                "last 2 Safari versions",
                "last 2 iOS versions",
                "last 2 Edge versions",
                "last 2 Opera versions",
                "not and_uc 15.5",
              ],
              "info": [Function],
              "options": {
                "overrideBrowserslist": [
                  "> 1%",
                  "last 1 Android versions",
                  "last 1 ChromeAndroid versions",
                  "last 2 Chrome versions",
                  "last 2 Firefox versions",
                  "last 2 Safari versions",
                  "last 2 iOS versions",
                  "last 2 Edge versions",
                  "last 2 Opera versions",
                  "not and_uc 15.5",
                ],
              },
              "postcssPlugin": "autoprefixer",
              "prepare": [Function],
            },
            {
              "RuleExit": [Function],
              "postcssPlugin": "postcss-progressive-custom-properties",
            },
            {
              "OnceExit": [Function],
              "postcssPlugin": "postcss-preset-env",
            },
          ],
          "postcssPlugin": "postcss-preset-env",
        },
        [Function],
        [Function],
        {
          "OnceExit": [Function],
          "postcssPlugin": "postcss-hash",
        },
        [Function],
      ],
    },
  },
}
`;
